feedback/cached.py

```diff
@@ -139,3 +148,78 @@ def get_obj(self, key, spec_getter, i18n_getter): # pylint: disable=unused-argum
 
 
 CachedForm = CachedForm(timeout=60*60)
+
+
+class MiscCache:
+    """Cache for storing miscellaneous content related to a course."""
+    def __init__(self, prefix=None, timeout=None) -> None:
+        self.prefix = prefix or self.__class__.__name__
+        self.timeout = timeout or 60 * 60
+
+    def get_suffix(self, *args) -> str:
+        return '-'.join(str(x) for x in args)
+
+    def get(self, key: str, course: Course):
+        full_key = '/'.join((self.prefix, self.get_suffix(key, course.id)))
+        return cache.get(full_key)
+
+    def set(self, key: str, course: Course, value, timeout=-1) -> None:
+        full_key = '/'.join((self.prefix, self.get_suffix(key, course.id)))
+        timeout = timeout if (timeout != -1) else self.timeout
+        cache.set(full_key, value, timeout)
+
+
+class BackgroundCache(MiscCache):
+    """Cache for storing information about background questionnaires.
+    Stores both background questionnaire (enrollment exercise) api ids with
+    with relevant information (such as question types, display texts, etc.)
+    as well as student responses to questionnaires.
+    """
+    def get_bg_questionnaires(self, course: Course)-> Optional[dict[int, dict]]:
+        return self.get('questionnaires', course)
+
+    def get_or_set_bg_questionnaires(self,
+            course: Course,
+            client: AplusTokenClient
+            ) -> dict[int, dict]:
+        bgq_dict = self.get_bg_questionnaires(course)
+        if bgq_dict is None: # not saved in cache yet
+            bgq_dict = get_bg_questionnaires(course, client)
+            # set in cache
+            self.set('questionnaires', course, bgq_dict)
+        return bgq_dict
+
+    def get_response(self, student: Student, course: Course) -> Optional[tuple[int, dict]]:
+        full_key = '/'.join((
+            self.prefix,
+            self.get_suffix(student.id, course.id, 'response'))
+        )
+        return cache.get(full_key)
+
+    def get_or_set_response(self,
+            student: Student,
+            course: Course,
+            client: AplusTokenClient
+            ) -> tuple[int, dict]:
+        response = self.get_response(student, course)
+        if response is None: # not saved in cache yet
+            bg_questionnaires = self.get_or_set_bg_questionnaires(course, client)
+            if len(bg_questionnaires) == 0:
+                # no background questionnaires on course
+                response = (None, None)
+            else:
+                response = get_student_bg_responses(
+                    student, client, bg_questionnaires
+                )
+            # set value in cache
+            full_key = '/'.join((
+                self.prefix,
+                self.get_suffix(student.id, course.id, 'response'))
+            )
```

Comment:
I'd prefer the closing parenthesis of the tuple to be on the same line as the other closing parenthesis (with an extra comma too), like this:
```
full_key = '/'.join((
    self.prefix,
    self.get_suffix(student.id, course.id, 'response'),
))
```